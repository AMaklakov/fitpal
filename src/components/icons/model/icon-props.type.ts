import { StyleProp, TextStyle, ViewStyle } from 'react-native';

export interface IconProps {
	onPress?: () => void;
	onLongPress?: () => void;
	onPressIn?: () => void;

	size?: number;
	color?: string;

	activeOpacity?: number;
	style?: StyleProp<TextStyle>;
	touchableStyle?: StyleProp<ViewStyle>;
}
