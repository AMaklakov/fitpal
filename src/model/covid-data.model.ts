import moment from 'moment';

export interface ICovidData {
	date: moment.Moment;
	confirmed: number;
	deaths: number;
	recovered: number;
}
