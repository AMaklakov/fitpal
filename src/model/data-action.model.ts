import { Action } from 'redux';

export interface DataAction<Payload = any> extends Action<string> {
	type: string;
	payload: Payload;
}

export type DataActionCreator<T = any> = (data: T, ...args: any[]) => DataAction<T>;
