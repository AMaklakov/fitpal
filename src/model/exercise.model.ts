import { assertUnreachable } from '../util/assert-unreachable';
import { isPresent } from '../util/type.util';

export enum ExerciseTypes {
	/**
	 * With regular weight
	 */
	Default,

	/**
	 * Exercise is being done with body weight, but additionally
	 * one can use some weight to make exercise `harder` to do
	 */
	WithAdditionalWeight,

	/**
	 * Exercise is being done with body weight, but additionally
	 * one can use some weight to make exercise `simpler` to do
	 */
	WithNegativeWeight,
}

export interface ExerciseModel {
	_id: string;

	type: ExerciseTypes;
	name: string;
}

export type ICreateExercise = Omit<ExerciseModel, '_id'>;

const isTypeValid = (type?: ExerciseTypes): boolean => {
	if (!isPresent(type)) {
		return false;
	}

	switch (type) {
		case ExerciseTypes.Default:
		case ExerciseTypes.WithAdditionalWeight:
		case ExerciseTypes.WithNegativeWeight:
			return true;
		default:
			assertUnreachable(type);
			return false;
	}
};

export const isExerciseValid = (exercise?: Partial<ExerciseModel>): boolean => {
	if (!isPresent(exercise)) {
		return false;
	}

	const { name, type } = exercise;

	if (!name) {
		return false;
	}

	if (!isPresent(type) || !isTypeValid(type)) {
		return false;
	}

	return true;
};
