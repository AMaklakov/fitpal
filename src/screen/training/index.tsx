import React from 'react';
import Training from './training';
import { TrainingScreenProps } from './types';
import { StoreModel } from '../../redux/store';
import { getCurrentDate } from '../../util/date.util';
import { getTrainingByDate } from '../../redux/selector/training.selector';
import { connect } from 'react-redux';
import { Routes } from '../navigator';
import { getExerciseList } from '../../redux/selector/exercise.selector';

const Screen = (props: TrainingScreenProps) => {
	const { training, navigation, exercises } = props;

	const addExerciseAction = () => {
		navigation.navigate({
			routeName: Routes.CreateTrainingExercise,
			params: { trainingId: training?.id },
		});
	};

	return (
		<Training
			training={training}
			canEdit={true}
			addExerciseAction={addExerciseAction}
			exercises={exercises}
		/>
	);
};

const mapStateToProps = (
	store: StoreModel
): Pick<TrainingScreenProps, 'training' | 'exercises'> => ({
	training: getTrainingByDate(store, getCurrentDate()),
	exercises: getExerciseList(store),
});

export default connect(mapStateToProps)(Screen);
