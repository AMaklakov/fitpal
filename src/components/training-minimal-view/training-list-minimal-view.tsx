import { TrainingModel } from '@model/training.model';
import { StyleSheet, View } from 'react-native';
import React from 'react';
import { TrainingMinimalView } from './training-minimal-view';
import { SwipeListView } from 'react-native-swipe-list-view';
import { Colors } from '../../css/colors.style';
import { SwipeHiddenButton } from '../swipe-list/button';

interface TrainingListMinimalViewProps {
	trainingList?: TrainingModel[];
	onTrainingPress?: (training: TrainingModel) => void;

	onCopy: (training: TrainingModel) => void;
	onDelete: (training: TrainingModel) => void;
}

const styles = StyleSheet.create({
	rowBack: {
		flex: 1,
		flexDirection: 'row',
		justifyContent: 'space-between',
	},
	button1: {
		flex: 1,
		alignItems: 'flex-start',
	},
	button2: {
		flex: 1,
		alignItems: 'flex-end',
	},
});

export const TrainingListMinimalView = (props: TrainingListMinimalViewProps) => {
	const { trainingList, onTrainingPress, onCopy = () => {}, onDelete = () => {} } = props;

	return (
		<SwipeListView<TrainingModel>
			data={trainingList}
			renderItem={(data, rowMap) => (
				<TrainingMinimalView training={data.item} onTrainingPress={onTrainingPress} />
			)}
			renderHiddenItem={(data, rowMap) => (
				<View style={styles.rowBack}>
					<SwipeHiddenButton
						style={styles.button1}
						title={'Copy'}
						item={data.item}
						onTouch={onCopy}
					/>

					<SwipeHiddenButton
						style={styles.button2}
						title={'Delete'}
						item={data.item}
						onTouch={onDelete}
						backgroundColor={Colors.LightRed}
						textColor={Colors.DarkRed}
					/>
				</View>
			)}
			leftOpenValue={75}
			stopLeftSwipe={100}
			rightOpenValue={-75}
			stopRightSwipe={-100}
		/>
	);
};
