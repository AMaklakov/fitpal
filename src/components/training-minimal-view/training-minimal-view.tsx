import React, { useCallback, useMemo, useRef } from 'react';
import { StyleSheet, TouchableOpacity, View } from 'react-native';
import { TrainingModel } from '@model/training.model';
import { H2 } from '@components/heading/h2';
import { Card, Divider, Tooltip } from 'react-native-elements';
import { useTranslation } from 'react-i18next';
import { CompactTrainingView } from '@screen/statistics/components/compact-training-view';
import { ExerciseModel } from '@model/exercise.model';
import { ITooltipMenuItem, TooltipMenu } from '@components/tooltip/tooltip-menu';

interface TrainingMinimalViewProps {
	training: TrainingModel;
	exercises: ExerciseModel[];
	onTrainingPress?: (training: TrainingModel) => void;
	onTrainingStart?: (training: TrainingModel) => void;
	onCopy?: (training: TrainingModel) => void;
	onDelete?: (training: TrainingModel) => void;
	onEdit?: (training: TrainingModel) => void;
	onExercisePress?: (exerciseId: string) => void;
}

export const TrainingMinimalView = (props: TrainingMinimalViewProps) => {
	const { training, exercises, onTrainingPress, onCopy, onDelete, onExercisePress, onTrainingStart, onEdit } = props;
	const { t } = useTranslation();
	const tooltip = useRef<Tooltip>(null);

	const handleTrainingPress = useCallback(() => onTrainingPress?.(training), [onTrainingPress, training]);
	const handleTrainingPlayPress = useCallback(() => {
		tooltip.current?.toggleTooltip();
		onTrainingStart?.(training);
	}, [onTrainingStart, training]);
	const handleDetailsPress = useCallback(() => {
		tooltip.current?.toggleTooltip();
		onTrainingPress?.(training);
	}, [onTrainingPress, training]);
	const handleOnCopy = useCallback(() => {
		tooltip.current?.toggleTooltip();
		onCopy?.(training);
	}, [onCopy, training]);
	const handleOnDelete = useCallback(() => {
		tooltip.current?.toggleTooltip();
		onDelete?.(training);
	}, [onDelete, training]);
	const handleEdit = useCallback(() => {
		tooltip.current?.toggleTooltip();
		onEdit?.(training);
	}, [onEdit, training]);

	const tooltipItems = useMemo<ITooltipMenuItem[]>(
		() => [
			{
				title: t('Details'),
				leftIcon: { name: 'search' },
				onPress: handleDetailsPress,
				key: 'details',
			},
			/* TODO check training exercise list and series */
			{
				title: t('Start training'),
				leftIcon: { name: 'play-arrow' },
				onPress: handleTrainingPlayPress,
				key: 'start',
				isShown: false,
			},
			{
				title: t('Edit'),
				leftIcon: { name: 'edit' },
				onPress: handleEdit,
				key: 'edit',
				isShown: !!onEdit,
			},
			{
				title: t('Copy'),
				leftIcon: { name: 'content-copy' },
				onPress: handleOnCopy,
				key: 'copy',
				isShown: !!onCopy,
			},
			{
				title: t('Delete'),
				leftIcon: { name: 'delete' },
				onPress: handleOnDelete,
				key: 'delete',
				isShown: !!onDelete,
			},
		],
		[handleDetailsPress, handleEdit, handleOnCopy, handleOnDelete, handleTrainingPlayPress, onCopy, onDelete, onEdit, t]
	);

	return (
		<Card
			title={
				<View style={styles.titleContainer}>
					<TouchableOpacity onPress={handleTrainingPress} style={styles.heading}>
						<H2 text={training?.name} />
					</TouchableOpacity>
					<TooltipMenu items={tooltipItems} tooltipRef={tooltip} />
				</View>
			}>
			<Divider />
			<CompactTrainingView
				useHeading={false}
				training={training}
				exercises={exercises}
				onExerciseNamePress={onExercisePress}
				onTotalPress={handleTrainingPress}
			/>
		</Card>
	);
};

const styles = StyleSheet.create({
	exerciseWrap: { maxWidth: 30 + '%' },
	heading: { flex: 1 },
	titleContainer: {
		marginBottom: 10,
		flexDirection: 'row',
		justifyContent: 'space-between',
		alignItems: 'center',
	},
});
