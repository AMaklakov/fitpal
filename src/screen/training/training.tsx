import { TrainingProps } from './types';
import React, { useState } from 'react';
import { ScrollView, View } from 'react-native';
import { H1 } from '@components/heading/h1';
import { TrainingModel } from '@model/training.model';
import { ShowTraining } from './show-training';
import { ReorderTrainingExercise } from './reorder-training-exercises';
import { EditIcon } from '@icons/edit.icon';
import { SaveIcon } from '@icons/save.icon';

const Training = (props: TrainingProps) => {
	const { training, addExerciseAction, exercises, removeExercise, changeTraining, canEdit } = props;

	if (!training) {
		throw new Error(`No training present`);
	}

	const { date } = training;
	const [isEdit, changeIsEdit] = useState(false);

	const reorderExercises = (t: TrainingModel) => changeTraining(t);

	const handleEditButtonPress = () => changeIsEdit(prevState => !prevState);

	return (
		<ScrollView>
			<View
				style={{
					flexDirection: 'row',
					paddingHorizontal: '5%',
					justifyContent: 'space-around',
				}}>
				<H1 text={training?.name + ' ' + date} />

				{canEdit && isEdit && <SaveIcon onPress={handleEditButtonPress} />}
				{canEdit && !isEdit && <EditIcon onPress={handleEditButtonPress} />}
			</View>

			{isEdit ? (
				<ReorderTrainingExercise exercises={exercises} training={training} changeTraining={reorderExercises} />
			) : (
				<ShowTraining
					exercises={exercises}
					addExerciseAction={addExerciseAction}
					removeExercise={removeExercise}
					training={training}
				/>
			)}
		</ScrollView>
	);
};

export default Training;
