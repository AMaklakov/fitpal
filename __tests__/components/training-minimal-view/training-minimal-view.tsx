import React from 'react';
import renderer from 'react-test-renderer';
import { TrainingMinimalView } from '@components/training-minimal-view/training-minimal-view';
import { TrainingModel } from '@model/training.model';

describe('TrainingMinimalView', () => {
	it('should render properly', () => {
		const training: TrainingModel = {
			name: 'Long name of exercise that causes showing ellipsis',
			exerciseList: [],
			date: '01.01.1999',
			id: '1',
		};

		const tree = renderer.create(<TrainingMinimalView training={training} onTrainingPress={() => {}} />);
		expect(tree).toMatchSnapshot();
	});
});
