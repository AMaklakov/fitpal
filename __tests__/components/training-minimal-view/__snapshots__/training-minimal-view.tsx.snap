// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrainingMinimalView should render properly 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={null}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#83AEC8",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 20,
        "paddingVertical": 10,
      }
    }
  >
    <View
      style={
        Object {
          "maxWidth": "70%",
        }
      }
    >
      <View
        style={Object {}}
      >
        <Text
          numberOfLines={1}
          style={
            Object {
              "fontSize": 20,
              "textAlign": "center",
            }
          }
        >
          Long name of exercise that causes showing ellipsis
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "maxWidth": "30%",
        }
      }
    >
      <Text>
        0
         упражнений
      </Text>
    </View>
  </View>
</View>
`;
